README.md
[![Sensu Bonsai Asset](https://img.shields.io/badge/Bonsai-Download%20Me-brightgreen.svg?colorB=89C967&logo=sensu)](https://bonsai.sensu.io/assets/alasconnect/sensu-prometheus-alert-check)
![Go Test](https://github.com/alasconnect/sensu-prometheus-alert-check/workflows/Go%20Test/badge.svg)
![goreleaser](https://github.com/alasconnect/sensu-prometheus-alert-check/workflows/goreleaser/badge.svg)

# prometheus-alert-check

## Table of Contents
- [prometheus-alert-check](#prometheus-alert-check)
  - [Table of Contents](#table-of-contents)
  - [Overview](#overview)
  - [Usage examples](#usage-examples)
    - [Asset registration](#asset-registration)
    - [Check definition](#check-definition)
  - [Installation from source](#installation-from-source)
  - [Contributing](#contributing)

## Overview

The prometheus-alert-check is a [Sensu Check][6] that monitors alerts in Prometheus.

## Usage examples

The command allows for the following arguments:

  - `-u, --url, PROMETHEUS_URL`: The base path of the Prometheus API. Default: **http://127.0.0.1:9090/**
  - `-i, --insecure-skip-verify, PROMETHEUS_SKIP_VERIFY`: Skip TLS certificate verification (not recommended!).
  - `-T, --trusted-ca-file, PROMETHEUS_CACERT`: TLS CA certificate bundle in PEM format.
  - `-t, --timeout, PROMETHEUS_TIMEOUT`: The number of seconds the test should wait for a response form the host (Default: 15).
  - `-f, --firing`: If specified, the check will only look for firing alerts.
  - `-p, --pending`: If specified, the check will only look for pending alerts.
  - `-l, --label`: Filter alerts by labels using a RegEx. Can be specified more than once. E.g. `--label myname=\"(Alert1|Alert2|^$)\"` will match 'Alert1', 'Alert2', and alerts with no label called 'myname'.
  - `-a, --annotation`: Filter alerts by annotations using a RegEx. Can be specified more than once.
  - `-w, --warning`: If specified, a failed check will result in a warning result.
  - `-c, --critical`: If specified, a failed check will result in a critical result. **Default**
  - `-F, --failure-level-label`: If specified, the result of a failed task will be determined by the specified Prometheus label. The label must have the value of 'warning' or 'critical'.
  - `-v, --verbose`: If specified, output will be more verbose.

### Asset registration

[Sensu Assets][10] are the best way to make use of this plugin. If you're not using an asset, please
consider doing so! If you're using sensuctl 5.13 with Sensu Backend 5.13 or later, you can use the
following command to add the asset:

```
sensuctl asset add alasconnect/sensu-prometheus-alert-check
```

If you're using an earlier version of sensuctl, you can find the asset on the [Bonsai Asset Index](https://bonsai.sensu.io/assets/alasconnect/sensu-prometheus-alert-check).

### Check definition

```yml
---
type: CheckConfig
api_version: core/v2
metadata:
  name: sensu-prometheus-alert-check
  namespace: default
spec:
  command: sensu-prometheus-alert-check --url http://localhost:9090/
  subscriptions:
  - system
  runtime_assets:
  - alasconnect/sensu-prometheus-alert-check
```

## Installation from source

The preferred way of installing and deploying this plugin is to use it as an Asset. If you would
like to compile and install the plugin from source or contribute to it, download the latest version
or create an executable script from this source.

From the local path of the sensu-prometheus-alert-check repository:

```
go build
```

## Contributing

For more information about contributing to this plugin, see [Contributing][1].

[1]: https://github.com/sensu/sensu-go/blob/master/CONTRIBUTING.md
[2]: https://github.com/sensu/sensu-plugin-sdk
[3]: https://github.com/sensu-plugins/community/blob/master/PLUGIN_STYLEGUIDE.md
[4]: https://github.com/alasconnect/sensu-prometheus-alert-check/blob/master/.github/workflows/release.yml
[5]: https://github.com/alasconnect/sensu-prometheus-alert-check/actions
[6]: https://docs.sensu.io/sensu-go/latest/reference/checks/
[7]: https://github.com/sensu/check-plugin-template/blob/master/main.go
[8]: https://bonsai.sensu.io/
[9]: https://github.com/sensu/sensu-plugin-tool
[10]: https://docs.sensu.io/sensu-go/latest/reference/assets/


CHANGELOG.md
# Changelog
All notable changes to this project will be documented in this file.

The format is based on [Keep a Changelog](http://keepachangelog.com/en/1.0.0/)
and this project adheres to [Semantic
Versioning](http://semver.org/spec/v2.0.0.html).

## Unreleased

## [0.1.0] - 2023-06-28

### Added
- Initial release


main.go
package main

import (
	"crypto/tls"
	"crypto/x509"
	"encoding/json"
	"fmt"
	"io"
	"io/ioutil"
	"log"
	"net/http"
	"net/url"
	"os"
	"regexp"
	"strings"
	"time"

	corev2 "github.com/sensu/core/v2"
	"github.com/sensu/sensu-plugin-sdk/sensu"
)

// Config represents the check plugin config.
type Config struct {
	sensu.PluginConfig
	PrometheusUrl      string
	InsecureSkipVerify bool
	TrustedCAFile      string
	Timeout            int
	FiringAlerts       bool
	PendingAlerts      bool
	Labels             map[string]string
	Annotations        map[string]string
	Warning            bool
	Critical           bool
	FailureLevelLabel  string
	VerboseLogging     bool
}

var (
	plugin = Config{
		PluginConfig: sensu.PluginConfig{
			Name:     "sensu-prometheus-alert-check",
			Short:    "A Sensu check for monitoring alerts in Prometheus.",
			Keyspace: "sensu.io/plugins/sensu-prometheus-alert-check/config",
		},
	}

	options = []sensu.ConfigOption{
		&sensu.PluginConfigOption[string]{
			Path:      "prometheus-url",
			Env:       "PROMETHEUS_URL",
			Argument:  "url",
			Shorthand: "u",
			Default:   "http://127.0.0.1:9090/",
			Usage:     "The base path of the Prometheus API.",
			Value:     &plugin.PrometheusUrl,
		},
		&sensu.PluginConfigOption[bool]{
			Path:      "insecure-skip-verify",
			Env:       "PROMETHEUS_SKIP_VERIFY",
			Argument:  "insecure-skip-verify",
			Shorthand: "i",
			Default:   false,
			Usage:     "Skip TLS certificate verification (not recommended!).",
			Value:     &plugin.InsecureSkipVerify,
		},
		&sensu.PluginConfigOption[string]{
			Path:      "trusted-ca-file",
			Env:       "PROMETHEUS_CACERT",
			Argument:  "trusted-ca-file",
			Shorthand: "T",
			Default:   "",
			Usage:     "TLS CA certificate bundle in PEM format.",
			Value:     &plugin.TrustedCAFile,
		},
		&sensu.PluginConfigOption[int]{
			Path:      "timeout",
			Env:       "PROMETHEUS_TIMEOUT",
			Argument:  "timeout",
			Shorthand: "t",
			Default:   15,
			Usage:     "The number of seconds the test should wait for a response form the host (Default: 15).",
			Value:     &plugin.Timeout,
		},
		&sensu.PluginConfigOption[bool]{
			Path:      "firing",
			Env:       "",
			Argument:  "firing",
			Shorthand: "f",
			Default:   false,
			Usage:     "If specified, the check will only look for firing alerts.",
			Value:     &plugin.FiringAlerts,
		},
		&sensu.PluginConfigOption[bool]{
			Path:      "pending",
			Env:       "",
			Argument:  "pending",
			Shorthand: "p",
			Default:   false,
			Usage:     "If specified, the check will only look for pending alerts.",
			Value:     &plugin.PendingAlerts,
		},
		&sensu.MapPluginConfigOption[string]{
			Path:      "labels",
			Env:       "",
			Argument:  "label",
			Shorthand: "l",
			Default:   make(map[string]string, 0),
			Usage:     "Filter alerts by labels using a RegEx. Can be specified more than once. E.g. '--label myname=\"(Alert1|Alert2|^$)\"' will match 'Alert1', 'Alert2', and alerts with no label called 'myname'.",
			Value:     &plugin.Labels,
		},
		&sensu.MapPluginConfigOption[string]{
			Path:      "annotations",
			Env:       "",
			Argument:  "annotation",
			Shorthand: "a",
			Default:   make(map[string]string, 0),
			Usage:     "Filter alerts by annotations using a RegEx. Can be specified more than once.",
			Value:     &plugin.Annotations,
		},
		&sensu.PluginConfigOption[bool]{
			Path:      "warning",
			Env:       "",
			Argument:  "warning",
			Shorthand: "w",
			Default:   false,
			Usage:     "If specified, a failed check will result in a warning result.",
			Value:     &plugin.Warning,
		},
		&sensu.PluginConfigOption[bool]{
			Path:      "critical",
			Env:       "",
			Argument:  "critical",
			Shorthand: "c",
			Default:   false,
			Usage:     "If specified, a failed check will result in a critical result. (Default)",
			Value:     &plugin.Critical,
		},
		&sensu.PluginConfigOption[string]{
			Path:      "failure-level-label",
			Env:       "",
			Argument:  "failure-level-label",
			Shorthand: "F",
			Default:   "",
			Usage:     "If specified, the result of a failed task will be determined by the specified Prometheus label. The label must have the value of 'warning' or 'critical'.",
			Value:     &plugin.FailureLevelLabel,
		},
		&sensu.PluginConfigOption[bool]{
			Path:      "verbose",
			Env:       "",
			Argument:  "verbose",
			Shorthand: "v",
			Default:   false,
			Usage:     "If specified, output will be more verbose.",
			Value:     &plugin.VerboseLogging,
		},
	}
)

func main() {
	useStdin := false
	fi, err := os.Stdin.Stat()
	if err != nil {
		fmt.Printf("Error check stdin: %v\n", err)
		panic(err)
	}
	//Check the Mode bitmask for Named Pipe to indicate stdin is connected
	if fi.Mode()&os.ModeNamedPipe != 0 {
		log.Println("using stdin")
		useStdin = true
	}

	check := sensu.NewGoCheck(&plugin.PluginConfig, options, checkArgs, executeCheck, useStdin)
	check.Execute()
}

func checkArgs(event *corev2.Event) (int, error) {
	if len(plugin.PrometheusUrl) == 0 {
		return sensu.CheckStateWarning, fmt.Errorf("--url or PROMETHEUS_URL environment variable is required")
	}

	_, err := url.Parse(plugin.PrometheusUrl)
	if err != nil {
		return sensu.CheckStateCritical, fmt.Errorf("failed to parse prometheus URL %s: %v", plugin.PrometheusUrl, err)
	}

	if plugin.PendingAlerts && plugin.FiringAlerts {
		return sensu.CheckStateWarning, fmt.Errorf("both --pending and --firing cannot be specified at the same time")
	}

	if plugin.Warning && plugin.Critical {
		return sensu.CheckStateWarning, fmt.Errorf("both --warning and --critical cannot be specified at the same time")
	}

	if len(plugin.FailureLevelLabel) > 0 && (plugin.Warning || plugin.Critical) {
		return sensu.CheckStateWarning, fmt.Errorf("--warning and --critical cannot be specified with --failure-level-label")
	}

	return sensu.CheckStateOK, nil
}

func executeCheck(event *corev2.Event) (int, error) {
	log.Printf("Executing check with --url \"%s\".\n", plugin.PrometheusUrl)

	baseUrl, err := url.Parse(plugin.PrometheusUrl)
	if err != nil {
		return sensu.CheckStateCritical, fmt.Errorf("failed to parse base URL %s: %v", plugin.PrometheusUrl, err)
	}

	client, err := GetHttpClient(baseUrl)
	if err != nil {
		return sensu.CheckStateCritical, fmt.Errorf("unable to build HTTP client: %v", err)
	}

	log.Println("Fetching alerts...")
	alerts, err := GetAlerts(client, baseUrl)
	if err != nil {
		return sensu.CheckStateCritical, fmt.Errorf("unable to fetch alerts: %v", err)
	}

	alerts, err = FilterAlerts(alerts)
	if err != nil {
		return sensu.CheckStateCritical, fmt.Errorf("unable filter alerts: %v", err)
	}

	if len(alerts) > 0 {
		log.Printf("%d FOUND ALERTS:\n", len(alerts))

		bytes, err := json.MarshalIndent(alerts, "", "  ")
		if err != nil {
			return sensu.CheckStateCritical, fmt.Errorf("unable to marshal alerts: %v", err)
		}

		fmt.Println(string(bytes))
		return GetFailureLevel(alerts), nil
	}

	log.Println("Check passed, no alerts found.")
	return sensu.CheckStateOK, nil
}

func GetFailureLevel(alerts []Alert) int {
	if len(plugin.FailureLevelLabel) > 0 {
		found := false
		result := sensu.CheckStateUnknown

		for _, alert := range alerts {
			label, ok := alert.Labels[plugin.FailureLevelLabel]
			if ok {
				found = true
				clean := strings.ToLower(strings.TrimSpace(label))

				// Pick the highest warning level.
				// If the result is not unknown, it must be warning or critical. We only want to override the unknown level.
				if clean == "warning" && result == sensu.CheckStateUnknown {
					result = sensu.CheckStateWarning
				} else if clean == "critical" { // Critical is the highest.
					result = sensu.CheckStateCritical
				}
			}
		}

		if !found {
			log.Println("Unable to find failure level label. Defaulting to unknown.")
			result = sensu.CheckStateUnknown
		}

		return result
	} else if plugin.Critical {
		return sensu.CheckStateCritical
	} else if plugin.Warning {
		return sensu.CheckStateWarning
	} else {
		// Use critical by default.
		return sensu.CheckStateCritical
	}
}

func CopyToEvent(names []string, src map[string]string, dest map[string]string) {
	for _, name := range names {
		value, ok := src[name]
		if ok {
			dest[name] = value
		}
	}
}

func CompileFilters(exps map[string]string) (map[string]regexp.Regexp, error) {
	result := make(map[string]regexp.Regexp, len(exps))

	for name, regex := range exps {
		LogVerbosef("Compiling filter %s=%s\n", name, regex)
		xp, err := regexp.Compile(regex)

		if err != nil {
			return nil, fmt.Errorf("failed to compile regex '%s': %v", regex, err)
		}

		result[name] = *xp
	}

	return result, nil
}

func MatchProperty(values map[string]string, filters map[string]regexp.Regexp) (bool, *string, *string) {
	result := true

	for name, regex := range filters {
		value, ok := values[name]

		if ok {
			result = regex.MatchString(value)
		} else {
			result = regex.MatchString("")
		}

		if !result {
			return result, &name, &value
		}
	}

	return result, nil, nil
}

func FilterAlerts(alerts []Alert) ([]Alert, error) {
	labelFilters, err := CompileFilters(plugin.Labels)
	if err != nil {
		return nil, fmt.Errorf("failed to compile label filter: %v", err)
	}

	annotationFilters, err := CompileFilters(plugin.Annotations)
	if err != nil {
		return nil, fmt.Errorf("failed to compile label filter: %v", err)
	}

	result := []Alert{}

	for _, alert := range alerts {
		match := true
		var name *string
		var value *string

		match = match && (!plugin.FiringAlerts || alert.State == "firing")
		match = match && (!plugin.PendingAlerts || alert.State == "pending")

		if !match {
			LogVerbosef("Alert ignored due to state: %s.\n", alert.State)
		}

		if match {
			match, name, value = MatchProperty(alert.Labels, labelFilters)

			if !match {
				LogVerbosef("Alert ignored due to label: %s=%s.\n", *name, *value)
			}
		}

		if match {
			match, name, value = MatchProperty(alert.Annotations, annotationFilters)

			if !match {
				LogVerbosef("Alert ignored due to annotation: %s=%s.\n", *name, *value)
				break
			}
		}

		if match {
			result = append(result, alert)
		}
	}

	return result, nil
}

func GetAlerts(client *http.Client, baseUrl *url.URL) ([]Alert, error) {
	url := baseUrl.String() + "/api/v1/alerts"
	log.Printf("Executing request \"GET %s\".\n", url)
	req, err := http.NewRequest(http.MethodGet, url, nil)

	if err != nil {
		return nil, fmt.Errorf("failed to create request: %v", err)
	}

	resp, err := client.Do(req)

	if err != nil {
		return nil, fmt.Errorf("failed to create request: %v", err)
	}

	bytes, err := io.ReadAll(resp.Body)

	if err != nil {
		return nil, fmt.Errorf("failed to read body of response: %v", err)
	}

	var data Alerts
	err = json.Unmarshal(bytes, &data)

	if err != nil {
		return nil, fmt.Errorf("failed to unmarshal response: %v", err)
	}

	return data.Data.Alerts, nil
}

type Alert struct {
	State       string            `json:"state"`
	ActiveAt    string            `json:"activeAt"`
	Value       string            `json:"value"`
	Labels      map[string]string `json:"labels"`
	Annotations map[string]string `json:"annotations"`
}

type Alerts struct {
	Status string `json:"status"`
	Data   struct {
		Alerts []Alert `json:"alerts"`
	} `json:"data"`
}

func GetHttpClient(url *url.URL) (*http.Client, error) {
	client := http.DefaultClient
	client.Timeout = time.Duration(plugin.Timeout) * time.Second
	tr := http.DefaultTransport

	if url.Scheme == "https" {
		var caCertPool *x509.CertPool = nil

		if len(plugin.TrustedCAFile) > 0 {
			caCert, err := ioutil.ReadFile(plugin.TrustedCAFile)
			if err != nil {
				return nil, fmt.Errorf("unable to read cert file at %s: %v", plugin.TrustedCAFile, err)
			}
			caCertPool = x509.NewCertPool()
			caCertPool.AppendCertsFromPEM(caCert)
		}

		tr = &http.Transport{
			TLSClientConfig: &tls.Config{
				InsecureSkipVerify: plugin.InsecureSkipVerify,
				RootCAs:            caCertPool,
			},
		}
	}

	client.Transport = tr
	return client, nil
}

func LogVerbosef(format string, v ...any) {
	if plugin.VerboseLogging {
		log.Printf(format, v...)
	}
}

func LogVerboseln(v ...any) {
	if plugin.VerboseLogging {
		log.Println(v...)
	}
}
